bson/bsonrw/value_writer.go

```diff
@@ -247,7 +248,12 @@ func (vw *valueWriter) invalidTransitionError(destination mode, name string, mod
 func (vw *valueWriter) writeElementHeader(t bsontype.Type, destination mode, callerName string, addmodes ...mode) error {
 	switch vw.stack[vw.frame].mode {
 	case mElement:
-		vw.buf = bsoncore.AppendHeader(vw.buf, t, vw.stack[vw.frame].key)
+		key := vw.stack[vw.frame].key
+		if !isValidCString(key) {
+			return errors.New("BSON element key cannot contain null bytes")
+		}
```

Comment:
Do you suspect this will have any impact on marshalling performance? I suspect not, but if you do, perhaps it's worth running the performance task on evergreen before merging.

Reply:
I did some testing locally with docs of size ~1000 with ~100 character key names and did not see a drastic perf decrease. I also tried special-casing marshalling to append one char at a time for key names and check for null bytes that way and didn't see any better perf, so I don't think the small hit is avoidable.

Reply:
Oh nice, thanks for investigating it locally!